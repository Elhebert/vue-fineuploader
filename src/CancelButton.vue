<template>
  <renderless-cancel-button
    :id="id"
    :uploader="uploader"
  >
    <button
      v-if="!onlyRenderIfCancelable || cancelable"
      slot-scope="{ buttonEvents, cancelable }"
      :disabled="!cancelable"
      type="button"
      v-on="buttonEvents"
    >
      Cancel
    </button>
  </renderless-cancel-button>
</template>

<script>
import RenderlessCancelButton from './renderless/CancelButton'

export default {
  components: {
    RenderlessCancelButton,
  },

  props: {
    id: {
      type: Number,
      required: true,
    },
    onlyRenderIfCancelable: {
      type: Boolean,
      default: true,
    },
    uploader: {
      type: Object,
      required: true,
    },
  },
}
</script>
