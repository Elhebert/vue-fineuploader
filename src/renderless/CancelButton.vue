<script>
export default {
  props: {
    id: {
      type: Number,
      required: true,
    },
    uploader: {
      type: Object,
      required: true,
    },
  },

  data: () => ({
    cancelable: true,
    unmounted: false,
    statuses: null,
    initialStatus: null,
  }),

  created() {
    this.statuses = this.uploader.qq.status
    this.initialStatus = this.uploader.methods.getUploads({
      id: this.id,
    }).status
    this.cancelable = this.isCancelableStatus(this.initialStatus)
  },

  mounted() {
    this.uploader.on('statusChange', this.onStatusChange)
  },

  beforeDestroy() {
    this.unmounted = true
    this.unregisterStatusChangeHandler()
  },

  methods: {
    onStatusChange(id, oldStatus, newStatus) {
      if (id !== this.id || this.unmounted) {
        return
      }

      if (this.isCancelableStatus(newStatus) === !this.cancelable) {
        this.cancelable = this.isCancelableStatus(newStatus)

        return
      }

      if (
        newStatus === this.statuses.DELETED ||
        newStatus === this.statuses.CANCELED
      ) {
        this.unregisterStatusChangeHandler()
      }
    },

    onClick(event) {
      event.preventDefault()

      if (this.cancelable) {
        this.uploader.methods.cancel(this.id)
      }
    },

    unregisterStatusChangeHandler() {
      this.uploader.off('statusChange', this.onStatusChange)
    },

    isCancelableStatus(status) {
      return (
        [
          this.statuses.DELETE_FAILED,
          this.statuses.PAUSED,
          this.statuses.QUEUED,
          this.statuses.RETRYING_UPLOAD,
          this.statuses.SUBMITTED,
          this.statuses.UPLOADING,
          this.statuses.UPLOAD_FAILED,
        ].indexOf(status) >= 0
      )
    },
  },

  render() {
    return this.$scopedSlots.default({
      buttonEvents: {
        click: event => {
          this.onClick(event)
        },
      },
      cancelable: this.cancelable,
    })
  },
}
</script>
