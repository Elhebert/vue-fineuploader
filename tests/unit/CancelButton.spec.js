import { mount } from '@vue/test-utils'
import CancelButton from '@/CancelButton.vue'
import FineUploader from 'fine-uploader-wrappers'
import { status as STATUSES } from 'fine-uploader/lib/core/all'

describe('CancelButton.vue', () => {
  let wrapper
  let uploader

  const sampleBlob = new Blob(['hi!'], { type: 'text/plain' })

  beforeEach(done => {
    uploader = new FineUploader({ options: { autoUpload: false } })
    uploader.on('submitted', done)
    uploader.methods.addFiles(sampleBlob)

    wrapper = mount(CancelButton, {
      propsData: { id: 0, uploader },
    })
  })

  it('renders the button for a submitted file', () => {
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.html()).toBeDefined()
    expect(wrapper.attributes().disabled).toBeUndefined()
  })

  it('cancels the upload if clicked', done => {
    wrapper.trigger('click')

    wrapper.vm.$nextTick(() => {
      expect(uploader.methods.getUploads()[0].status).toMatch(STATUSES.CANCELED)

      done()
    })
  })

  it('removes the button by default if the file can no longer be canceled', done => {
    uploader.methods.cancel(0)

    wrapper.vm.$nextTick(() => {
      expect(wrapper).toMatchSnapshot()
      expect(wrapper.isEmpty()).toBe(true)

      done()
    })
  })

  it('disables the button if requested when the file can no longer be canceled', done => {
    wrapper.setProps({ onlyRenderIfCancelable: false })
    uploader.methods.cancel(0)

    wrapper.vm.$nextTick(() => {
      expect(wrapper).toMatchSnapshot()
      expect(wrapper.html()).toBeDefined()
      expect(wrapper.attributes().disabled).toBeDefined()

      done()
    })
  })
})
