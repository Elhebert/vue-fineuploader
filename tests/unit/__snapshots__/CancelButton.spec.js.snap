// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CancelButton.vue disables the button if requested when the file can no longer be canceled 1`] = `
<button type="button" disabled="disabled">
  Cancel
</button>
`;

exports[`CancelButton.vue removes the button by default if the file can no longer be canceled 1`] = ``;

exports[`CancelButton.vue renders the button for a submitted file 1`] = `
<button type="button">
  Cancel
</button>
`;
